File: src/main/java/org/dukcode/ps/codetree/trail02/chapter01/lesson02/IntroAnEvenNumberWithAMultipleOf5InTheSum.java
package org.dukcode.ps.codetree.trail02.chapter01.lesson02;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;

/**
 * @see <a
 * href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-an-even-number-with-a-multiple-of-5-in-the-sum/description">LINK</a>
 */
public class IntroAnEvenNumberWithAMultipleOf5InTheSum {

  private static BufferedReader br;
  private static BufferedWriter bw;

  public static void main(String[] args) throws IOException {
    br = new BufferedReader(new InputStreamReader(System.in));
    bw = new BufferedWriter(new OutputStreamWriter(System.out));

    int n = Integer.parseInt(br.readLine());

    bw.write(solve(n) ? "Yes" : "No");

    br.close();
    bw.close();

  }

  private static boolean solve(int n) {
    if (n % 2 != 0) {
      return false;
    }

    int digitSum = 0;
    while (n > 0) {
      digitSum += n % 10;
      n /= 10;
    }

    return digitSum % 5 == 0;
  }
}

File: src/main/kotlin/org/dukcode/ps/codetree/trail02/chapter01/lesson02/IntroAnEvenNumberWithAMultipleOf5InTheSum.kt
package org.dukcode.ps.codetree.trail02.chapter01.lesson02

import java.io.BufferedReader
import java.io.BufferedWriter
import java.io.InputStreamReader
import java.io.OutputStreamWriter

private val br = BufferedReader(InputStreamReader(System.`in`))
private val bw = BufferedWriter(OutputStreamWriter(System.out))

/**
 * @see <a href="https://www.codetree.ai/ko/trails/complete/curated-cards/intro-an-even-number-with-a-multiple-of-5-in-the-sum/description">LINK</a>
 */
fun main() {
    val num = br.readLine().toInt()
    bw.write(if (isMagicNumber(num)) "Yes" else "No")

    bw.flush()
}

private fun isMagicNumber(num: Int): Boolean = num % 2 == 0 && digitSum(num) % 5 == 0

private fun digitSum(num: Int): Int {
    var ret = 0

    var value = num
    while (value > 0) {
        ret += value % 10
        value /= 10
    }

    return ret
}

/*
fun digitSum2(num: Int): Int = generateSequence(num) { it / 10 }
        .takeWhile { it > 0 }
        .sumOf { it % 10 }
*/
